### PR TITLE
Add Condition and Reasons for VirtualMachineClassBindings and VirtualMachineClasses

### PR DESCRIPTION
**Pull request by aruneshpa**
There are many developer workflows where do not provide any user feedback. This
change defines a Condition and a few reasons to address some of those
workflows.

The following Condition is defined:
- VirtualMachinePreReqReady:
A condition that represents that all the Prerequisites of a `VirtualMachine` are
ready. For now, we only check for the availability of `VirtualMachineClass` and
the that the class is associated with the VirtualMachine's namespace. In future,
this Condition can be used to check for other components as well
(e.g. `VirtualMachineImage`, `StorageClass` etc).

A few reasons are also defined:
- VirtualMachineClassBindingNotFound
If the VirtualMachine's namespace does not have access to the specified
`VirtualMachineClass`.
- VirtualMachineClassNotFound
If the specified `VirtualMachineClass` does not exist on the cluster.

### CODE REVIEW

**Comment by janitha** [comment]
Prerequisite is a single word, so no need to uppercase the R in the middle

File: api/v1alpha1/condition_consts.go [hunk 1]
```diff
@@ -9,15 +9,33 @@ const (
 	ReadyCondition ConditionType = "Ready"
 )
 
+const (
+	// VirtualMachinePreReqReadyCondition is set on VirtualMachines when all of a VirtualMachine's prerequisites are satisfied.
+	// For now, this check only validates that the VirtualMachineClass specified in the VirtualMachineSpec exists on the API
+	// server and the VirtualMachine's namespace has access to the specified VirtualMachineClass.
+	VirtualMachinePreReqReadyCondition = "VirtualMachinePreReqReady"
+
+	// VirtualMachinePreReqReadyReason documents that all of a VirtualMachine's PreRequisites have been satisfied.
```
Comment (janitha) [inline api/v1alpha1/condition_consts.go:18]: Since the condition type already identifies that this is for VirtualMachinePreReqReady, seems redundant to also have the Reason be that verbose, it can just be "Ready" like other conditions.

**Review by janitha — APPROVED** [review]

File: api/v1alpha1/condition_consts.go [hunk 2]
```diff
@@ -9,15 +9,33 @@ const (
 	ReadyCondition ConditionType = "Ready"
 )
 
+const (
+	// VirtualMachinePreReqReadyCondition is set on VirtualMachines when all of a VirtualMachine's prerequisites are satisfied.
+	// For now, this check only validates that the VirtualMachineClass specified in the VirtualMachineSpec exists on the API
```
Comment (bryanv) [inline api/v1alpha1/condition_consts.go:14]: I don't think we need to call out "on the API server"; just exists is sufficient.

I'd reword this so we don't have to constantly update it, and for it to become out of date when we do. Say something along the lines of the reasons below are satisfied.

File: api/v1alpha1/condition_consts.go [hunk 3]
```diff
@@ -9,15 +9,33 @@ const (
 	ReadyCondition ConditionType = "Ready"
 )
 
+const (
```
[the remark this answers is not in the document]
Comment (aruneshpa) [reply api/v1alpha1/condition_consts.go:14]: Done.

**Comment by jainvarun21** [comment]
Would this condition be relevant only during VM create? Have we considered an alternate workflow wherein we do not let the user create the VM k8s resource if the pre-reqs are not met (e.g. via webhooks)?

**Comment by aruneshpa** [comment]
> Would this condition be relevant only during VM create?
@jainvarun21 , these conditions will be exposed in `Create` and `Update` VM paths. We want to expose these conditions as soon as possible so developers can take appropriate action (e.g. ask VI admins to add the VM class to a namespace).

>  Have we considered an alternate workflow wherein we do not let the user create the VM k8s resource if the pre-reqs are not met (e.g. via webhooks)?
Right now, we do enforce some of these constraints (checking that the image and class is specified). Since the VM class attach/detach is handles by wcpsvc asynchronously, it is perfectly fine for a VM to wait in the reconciler thread for the class binding to be available.

File: api/v1alpha1/condition_consts.go [hunk 4]
```diff
@@ -9,15 +9,34 @@ const (
...
 )
 
+// Conditions and condition Reasons for the VirtualMachine object.
+
+const (
+	// VirtualMachinePrereqReadyCondition documents that all of a VirtualMachine's prerequisites declared in the spec
+	// (e.g. VirtualMachineClass) are satisfied.
+	VirtualMachinePrereqReadyCondition ConditionType = "VirtualMachinePrereqReady"
+
+	// VirtualMachinePrereqReadyReason documents that all of a VirtualMachine's PreRequisites have been satisfied.
```
Comment (jainvarun21) [inline api/v1alpha1/condition_consts.go:19]: Do we need to set a Reason while the condition is true?

Comment (aruneshpa) [reply api/v1alpha1/condition_consts.go:19]: No.

https://github.com/kubernetes-sigs/cluster-api/blob/dadd7bd6233edb1e422e7af282e923abff52f830/util/conditions/setter.go#L109

Comment (jainvarun21) [reply api/v1alpha1/condition_consts.go:19]: In that case do we need VirtualMachinePrereqReadyReason?

Comment (aruneshpa) [reply api/v1alpha1/condition_consts.go]: Removed the `Reason` for the True condition. Thanks for pointing that out.
